tools/firmware-utils/src/alpha_encimg.c

```diff
@@ -0,0 +1,141 @@
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#include <openssl/aes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define TEMPFILE "alpha_encimg.tmp"
+
+void alpha_xor(unsigned char* input_bytes, unsigned char input_size,
+		unsigned char* xorkey, unsigned char xorkey_size, unsigned char* out)
+{
+	for(unsigned char i=0; i < input_size; i++)
+	{
+		out[i] = (i + 1) ^ input_bytes[i];
+		out[i] = xorkey[i % xorkey_size] ^ out[i];
+	}
+}
+
+int main(int argc, char **argv)
+{
+	unsigned char *seama_signature = argv[2];
+	unsigned char *key = argv[3];
+	unsigned char *iv = argv[4];
+
+	unsigned char readbuf[4096];
+	unsigned char encbuf[4096];
+	unsigned char aes_key[32];
+	unsigned char aes_iv[16];
+
+	unsigned int read_bytes = 0;
+	unsigned long read_total = 0;
+	unsigned int blocks = 0;
+
+	if(argc != 5)
+	{
+		fprintf(stderr, "Usage:\n"
+			"\talpha_encimg file aes_key iv seama_signature\n\n");
+		exit(1);
+	}
+
+	FILE * input_file = fopen(argv[1], "rb");
+	if(input_file == NULL)
+	{
+		fprintf(stderr, "File %s could not be opened.", argv[1]);
+		exit(1);
+	}
+
+	alpha_xor(key, 32, seama_signature, strlen(seama_signature), aes_key);
+	alpha_xor(iv,  16, seama_signature, strlen(seama_signature), aes_iv);
+
+	AES_KEY enc_key;
+	AES_set_encrypt_key(aes_key, 256, &enc_key);
+
+	FILE * output_file = fopen(TEMPFILE, "wb");
+
+	while(read_bytes = fread(&readbuf, 1, 4096, input_file))
+	{
+		read_total += read_bytes;
+		if(read_bytes == 4096)
+		{
+			AES_cbc_encrypt(readbuf, encbuf, 4096, &enc_key, aes_iv,
+				AES_ENCRYPT);
+			fwrite(&encbuf, 1, 4096, output_file);
+		}
+		else
+		{
+			blocks = read_bytes / 16;
+			for(unsigned char i = 0; i < blocks; i++)
+			{
+				AES_cbc_encrypt(&readbuf[i * 16], encbuf, 16, &enc_key,
+					aes_iv, AES_ENCRYPT);
+				fwrite(&encbuf, 1, 16, output_file);
+			}
+			break;
+		}
+	}
+
+	fclose(input_file);
+
+	if(read_bytes % 16)
+	{
+		// pad to 16 bytes
+		unsigned char padding = 16 - (read_bytes % 16);
+		memset(&readbuf[read_bytes], 0, padding);
+		AES_cbc_encrypt(&readbuf[blocks * 16], encbuf, 16, &enc_key, aes_iv,
+			AES_ENCRYPT);
+		fwrite(&encbuf, 1, 16, output_file);
+	}
+	else
+	{
+		// force padding 16 bytes if input is already aligned
+		memset(&readbuf, 0, 16);
+		AES_cbc_encrypt(readbuf, encbuf, 16, &enc_key, aes_iv, AES_ENCRYPT);
+		fwrite(&encbuf, 1, 16, output_file);
```

Comment:
Validate the input file has the minimum required length for performing these operations.

Reply:
While I agree it would be quite pointless for the user to supply a 0-bytes file (which should result in a 16 bytes file of encrypted padding), this could most probably happen as a subsequent error, and as such should indeed terminate the build.
Or could there even be further issues (e.g. memory leak) with this that I overlooked?